examples/04.MuonScan/Validate.C

```diff
@@ -1,103 +1,140 @@
+        cout << "The number of registered does not match the reference value of 2" << endl;
         return 7;
     }
 
-    TRestGeant4Event* ev = (TRestGeant4Event*)run->GetInputEvent();
-    run->GetEntry(9);
+    TRestGeant4Event* event = run.GetInputEvent<TRestGeant4Event>();
+
+    double nEvents = run.GetEntries();
+
+    double averageTotalEnergy = 0;
+    constexpr double averageTotalEnergyRef = 18.0321;
+
+    double averageSensitiveEnergy = 0;
+    constexpr double averageSensitiveEnergyRef = 8.78066;
+
+    double averageNumberOfTracks = 0;
+    constexpr double averageNumberOfTracksRef = 4.924;
+
+    double averageNumberOfHitsVolume0 = 0;
+    constexpr double averageNumberOfHitsVolume0Ref = 60.712;
 
-    cout << "Total energy : " << ev->GetTotalDepositedEnergy() << endl;
-    Int_t en = (Int_t)(100 * ev->GetTotalDepositedEnergy());
-    cout << "Energy integer : " << en << endl;
-    cout << "Sensitive volume energy : " << ev->GetSensitiveVolumeEnergy() << endl;
-    cout << "Number of hits : " << ev->GetNumberOfHits() << endl;
-    cout << "Number of tracks : " << ev->GetNumberOfTracks() << endl;
+    double averageNumberOfHitsVolume1 = 0;
+    constexpr double averageNumberOfHitsVolume1Ref = 68.061;
 
-    Int_t X = (Int_t)(100 * ev->GetMeanPositionInVolume(0).X());
-    Int_t Y = (Int_t)(100 * ev->GetMeanPositionInVolume(0).Y());
-    Int_t Z = (Int_t)(100 * ev->GetMeanPositionInVolume(0).Z());
+    TVector3 averagePosition = {};
+    const TVector3 averagePositionRef = {-0.0305454, -0.384008, -299.54};
 
-    cout << "x: " << X << " y: " << Y << " z: " << Z << endl;
+    constexpr double tolerance = 0.05;
 
-    if (en != 1206) {
-        cout << "Error in total energy" << endl;
+    for (size_t i = 0; i < run.GetEntries(); i++) {
+        run.GetEntry(i);
+
+        averageTotalEnergy += event->GetTotalDepositedEnergy() / nEvents;
+        averageSensitiveEnergy += event->GetSensitiveVolumeEnergy() / nEvents;
+        averageNumberOfHitsVolume0 += event->GetNumberOfHits(0) / nEvents;
+        averageNumberOfHitsVolume1 += event->GetNumberOfHits(1) / nEvents;
+        averageNumberOfTracks += event->GetNumberOfTracks() / nEvents;
+        averagePosition += event->GetMeanPositionInVolume(0) * (1.0 / nEvents);
+    }
+
+    cout << "Average total energy: " << averageTotalEnergy << " keV" << endl;
+    cout << "Average sensitive energy: " << averageSensitiveEnergy << " keV" << endl;
+    cout << "Average number of hits in volume 0: " << averageNumberOfHitsVolume0 << endl;
+    cout << "Average number of hits in volume 1: " << averageNumberOfHitsVolume1 << endl;
+    cout << "Average number of tracks: " << averageNumberOfTracks << endl;
+    cout << "Average position: (" << averagePosition.x() << ", " << averagePosition.y() << ", "
+         << averagePosition.z() << ") mm" << endl;
+
+    if (TMath::Abs(averageNumberOfTracks - averageNumberOfTracksRef) / averageNumberOfTracksRef >
+        tolerance * 5) {
```

Comment:
I don't understand why tolerance has been included. As far as I know if you use the same seed you should get exactly the same value.

Reply:
Fair enough

Reply:
Actually it feels bad to me changing geant4 version in the pipeline. This breaks consistency. Is it necessary to do that? If the latest version is needed, for example, to test some new code feature, I think it is better to insert an individual stage in the pipeline with a new docker image.

As for `tolerance`, I think we should not use it to much in the validation file. If some commits are being pushed influencing simulation result, by using `tolerance` we won't be able to notice them from pipeline.